### DISCUSSION TITLE
Moving forward

### DISCUSSION THREAD

**Discussion by anzz1** [ggml-org/llama.cpp#346] [Polls]
Do the original coding guidelines still apply?

> - Avoid adding third-party dependencies, extra files, extra headers, etc.
> - Always consider cross-compatibility with other operating systems and architectures
> - Avoid fancy looking modern STL constructs, use basic for loops, avoid templates, keep it simple

The original codebase was minimalistic, clean and fast C-style code.

However, recent changes have:
- Added C++17 as a requirement
- Started to lean heavily towards using "modern" STL type constructs
- Added lots of platform-specific stuff

I do understand that some people want to rather work with modern C++/STL instead of C and a lot can be done easily with external libraries. However, in my opinion, the C-style, platform-independent, minimalistic style is exactly the whole point which makes this project special.

So my question is, is this the direction the project is going to take?

If so, I've forked a minimalistic version here https://github.com/anzz1/llama.cpp-min with the aim of keeping in line with the original codebase, keeping C-style in favor of C++/STL and removing all platform-specifics.

**Comment by niansa** [comment]
Intermediate. Modern C++ to a point that does not affect performance (overhead free abstractions), avoidance of third party dependencies and platform-specific code where it improved performance on these platforms.

**Comment by Green-Sky** [comment]
as far as i can tell, it's already moving back to c++11 https://github.com/ggerganov/llama.cpp/pull/350

... I wonder why people seem so fast to fork ...
